### PR TITLE
Favor done over then for thenables that have it

### LINKED ISSUE #137: Favour `done()` over `then()` for thennables that have it.
`ApiTrait::dispatch()` checks for the existence of a `then` method to identify generic promises. However, if the promise object also has a `done` method, this should be used instead.

Otherwise assertion failures or any other unexpected exception that occurs while resuming the strand will get trapped in the promise chain instead of reported back to the calling coroutine.

### PR DESCRIPTION
**Pull request by koden-km**
(fixes #137)

### CODE REVIEW

**Comment by jmalloc** [comment]
I pushed another commit to try to force Travis CI to build, it worked this time. My guess is there was some issue with the webhooks queuing on GitHub's end. We can rebase those commits away once the feature is done.

File: test/suite-functional/api/functional.thenable.spec.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
+<?php
+
+declare (strict_types = 1); // @codeCoverageIgnore
+
+namespace Recoil;
+
+use Eloquent\Phony\Phony;
+
+context('can be invoked by yielding a thenable', function () {
```
Comment (jmalloc) [inline test/suite-functional/api/functional.thenable.spec.php]: In regards to these test names, try to make them read well including the word "it". Perhaps:

`context('when it has a then method', ...)`
`rit('resumes the strand when resolved`, ...)`
`rit('resumes the strand with an exception when rejected`, ...)`

etc

Comment (koden-km) [inline test/suite-functional/api/functional.thenable.spec.php]: What about the "fulfil" vs "resolved"? Different names seem to be used in different libraries, which one do you want in recoil?

Comment (jmalloc) [inline test/suite-functional/api/functional.thenable.spec.php]: I tend to use "resolve" and "reject".

File: test/suite-functional/api/functional.thenable.spec.php [hunk 2]
```diff
@@ -53,10 +53,10 @@
         }
     });
 
-    rit('that does cancel', function () {
+    rit('terminates the strand when cancelled', function () {
```
Comment (jmalloc) [inline test/suite-functional/api/functional.thenable.spec.php]: This is backwards - it cancels the promise when the strand is terminated :P

File: test/suite-functional/api/functional.thenable.spec.php [hunk 3]
```diff
@@ -36,10 +36,10 @@
         }
     });
 
-    rit('that does reject with non throwable', function () {
+    rit('resumes the strand with a value when rejected', function () {
```
Comment (jmalloc) [inline test/suite-functional/api/functional.thenable.spec.php]: This resumes the strand with an exception when the promise is rejected with a non-exception.

**Pull request merged**